### PR TITLE
expression.ts: rate-limit the maximum number of batches in flight

### PR DESCRIPTION
**Pull request by austinEng**
avoids sending too many pipeline compilations at once which may stall

Issue: #

<hr>

**Requirements for PR author:**

- [ ] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [ ] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [ ] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/expression.ts [hunk 1]
```diff
@@ -251,12 +251,32 @@ export async function run(
   // A cache to hold built shader pipelines.
   const pipelineCache = new Map<String, GPUComputePipeline>();
 
-  // Submit all the cases in batches, each in a separate error scope.
-  const checkResults: Array<Promise<void>> = [];
+  // Submit all the cases in batches, rate-limiting to ensure not too many
+  // batches are in flight simultaneously.
+  const maxBatchesInFlight = 5;
```
Comment (austinEng) [inline src/webgpu/shader/execution/expression/expression.ts:256]: this number was chosen experimentally based on making it as small as I could without impacting total wall time of the test

File: src/webgpu/shader/execution/expression/expression.ts [hunk 2]
```diff
@@ -251,12 +251,32 @@ export async function run(
...
+    if (resolvePromiseBlockingBatch) {
+      resolvePromiseBlockingBatch();
+      resolvePromiseBlockingBatch = undefined;
+    }
+  };
+
   for (let i = 0; i < cases.length; i += casesPerBatch) {
     const batchCases = cases.slice(i, Math.min(i + casesPerBatch, cases.length));
 
-    t.device.pushErrorScope('validation');
```
Comment (kainino0x) [inline src/webgpu/shader/execution/expression/expression.ts:259]: error scope deemed not necessary? (seems fine to me)
You know, it occurs to me... #2610

**Review by kainino0x — APPROVED** [review]
LGTM

File: src/webgpu/shader/execution/expression/expression.ts (hunk 2 above)
Comment (austinEng) [reply src/webgpu/shader/execution/expression/expression.ts:259]: Seems like it's intended to make the checks occur in a delayed manner, but it shouldn't ultimately be necessary.

**Pull request merged**